### PR TITLE
Frontend/manual dataset upload

### LINKED ISSUE #147: [Frontend] - Manual Upload of Datasets
As a user, I would like to manually upload a dataset through a form.

**Description:**
This describes having the option to upload a dataset manually by filling out a form with all the appropriate fields. This form should be found on the same page as the File upload page.

Field Validation:

For the variables users should be able to enter any variable value for any variable
Users should be able to add more variables
Users should be able to add more "points"
Users should be able to select the materials existing in the db or create a new one

Link to JSON template which includes all the necessary fields for the form (JsonSample.json) : https://drive.google.com/drive/folders/1JWyLKra_hNhUOp19ioMeHPkk8Mxnf-8N

**Tasks Involved:**

- [x]  Create a form that accepts each field in the database (Please refer to the E-R Diagram to gather all the requirements)
- [x]  UI Test

**Mockup Image or link to Mockup:**
To come

**Acceptance Criteria**
- Required fields must be filled
- Must give the opportunity to fill out each dataset detail (fill out all the tables in the backend)
- UI test must be there

**Story Point** - 3
**Risk** - Low
**Priority** - High

### PR DESCRIPTION
**Pull request by ds161833**
Created the following page:
image

The page allows to modify a dataset.
image
image
field validation exists on the form

Materials have a nice selector:
image

Authors can be added or removed:
image

DataTable:
image

Select rows and then remove them
image

Variable header:
image

### CODE REVIEW

File: Client/src/Components/DatasetUpload/DataSection/DatasetDataTable.tsx [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+import { EditVariableHeader } from './EditVariableHeader'
+
+interface IProps {
+  data: IData,
+  onDataChange: (newData: IData) => void
+}
+
+export const DatasetDataTable = (props: IProps) => {
+  const [editedVariableIndex, setEditedVariableIndex] = useState(-1)
+  const [selectedRows, setSelectedRows] = useState(() => new Set<React.Key>())
```
Comment (GD-Dheer) [inline Client/src/Components/DatasetUpload/DataSection/DatasetDataTable.tsx]: Whats going on here? Ive never seen new Set<React.key> before

File: Client/src/Components/DatasetUpload/DataSection/EditVariableHeader.tsx [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+  const handleClose = () => {
+    props.onEditModalClose()
+  }
+
+  const openEditVariableModal = () => {
+    props.onHeaderClick(props.index)
+  }
+
+  const formik = useFormik({
```
Comment (GD-Dheer) [inline Client/src/Components/DatasetUpload/DataSection/EditVariableHeader.tsx:32]: formik A+

File: Client/src/Components/DatasetUpload/DatasetUploadForm.tsx [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    initialValues: { meta, reference, data },
+    onSubmit: values => {
+      let dataset: IDatasetModel = { ...values.meta, reference: values.reference, data: values.data }
+      onSubmit(dataset)
+    },
+  })
+
+  return (
+    <>
+      <form onSubmit={formik.handleSubmit}>
```
Comment (GD-Dheer) [inline Client/src/Components/DatasetUpload/DatasetUploadForm.tsx:34]: Nicely done. Separated the the form into distinct components

**Review by GD-Dheer — APPROVED** [review]
Looks good to me. Well coded and (portions that needed to be reworked were already commented). UI tests were also added!

File: Client/src/Components/DatasetUpload/DataSection/DatasetDataTable.tsx (hunk 1 above)
Comment (ds161833) [reply Client/src/Components/DatasetUpload/DataSection/DatasetDataTable.tsx:17]: good question

Comment (ds161833) [reply Client/src/Components/DatasetUpload/DataSection/DatasetDataTable.tsx:17]: cleaned up, was a copy paste from react-data-grid docs

Comment (ds161833) [reply Client/src/Components/DatasetUpload/DataSection/DatasetDataTable.tsx]: new Set<React.Key>() its a set data structure of type react key, we use it here to specify a set of numbers that represents the rows that are currently selected

Comment (ds161833) [reply Client/src/Components/DatasetUpload/DataSection/DatasetDataTable.tsx]: concerning the **new** https://www.geeksforgeeks.org/javascript-new-keyword/#:~:text=New%20keyword%20in%20JavaScript%20is,new%20empty%20object%20is%20created.

**Review by Mehrdadmaskull — APPROVED** [review]
Beautiful!! LGTM

**Pull request merged**